okio/src/main/java/okio/BufferCursor.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio;
+
+/**
+ * Direct access to the underlying byte arrays in a buffer.
+ *
+ * Note that cursors are stateful and can be reused.
+ *
+ * It is an error to mutate a buffer that is currently being operated on with a cursor.
+ *
+ * Currently this cursor only supports reading. It is an error to mutate the fields of this cursor
+ * or the values of the data[] array.
+ *
+ * Access should be in acquire/release pairs. This guarded access may be used to detect bugs where
```

Comment:
How about `implement Closeable` and drop the checked exception? This lets you drop 'em in try-with-resources. We do this in SQL Brite to great effect.

Reply:
https://square.github.io/sqlbrite/3.x/sqlbrite/com/squareup/sqlbrite3/BriteDatabase.html#newTransaction--

Reply:
Only drawback of that is that it makes it weird to re-use the same cursor. I’m imagining tools like BrotliSink having a single cursor and reusing it so there’s no allocation for the frames of a stream.

Reply:
Though I think the try-with-resources is probably worth it.

Reply:
The trick SQL Brite uses to avoid that is there's only one instance of the returned object.

So
```java
Transaction t1, t2;
try (Transaction t = db.newTransaction()) {
  t1 = t;
}
try (Transaction t = db.newTransaction()) {
  t2 = t;
}
assertSame(t1, t2);
```
would pass.

It's purely an API scoping mechanism.

So here you're remove all public APIs from `BufferCursor` except `acquire` and put it on some nested type. `BufferCursor` instances would hold a single instance of that nested type, and `acquire` would mutate it and return.

```java
BufferCursor c = new BufferCursor();
try (Something s = c.acquire(buffer)) {
  s.data...
}
try (Something s = c.acquire(buffer)) {
  s.data...
}
```

You almost want to make the nested type called `Cursor` and give the outer type some other name. Or maybe just hang this directly on `Buffer`!

```java
try (Cursor c = buffer.acquireUnsafeCursor()) {
  c.data...
}
try (Cursor c = buffer.acquireUnsafeCursor()) {
  c.data...
}
```

Reply:
SQL Brite details:
  * Single internal instance: https://github.com/square/sqlbrite/blob/dae34762ca41f532094433c3ac1e515865e74754/sqlbrite/src/main/java/com/squareup/sqlbrite3/BriteDatabase.java#L73
 * Returned from `newTransaction()`: https://github.com/square/sqlbrite/blob/dae34762ca41f532094433c3ac1e515865e74754/sqlbrite/src/main/java/com/squareup/sqlbrite3/BriteDatabase.java#L235

Reply:
Is it an error to acquire multiple cursors on the same buffer?  Should a buffer know there is an acquired cursor so it can error reads and writes until it is released?  Hanging cursor acquisition off of `Buffer` would allow for some of these checks and also use of the cursor internally for the `findSegmentAndOffset` blocks.

Reply:
Good questions. I think I'm a "yes" and "yes" for them.

Reply:
I’m actually “no” for what I think is an interesting reason.

I like “effectively immutable” objects. The best example is `Collections.unmodifiableList()` to wrap a mutable type like `ArrayList`, but even mutable types can be documented “please do not mutate me” and provide a lot of value. For example, in `Base64.java` we use a `byte[]` as an effectively immutable value. In Java almost all of the good mutable types are safe for multiple concurrent reads as long as the reference to the instance itself is safely published.

It’s not unreasonable to use `Buffer` as an effectively-immutable object. Perhaps you populate it with a bunch of data (a bloom filter? some OpenGL geometry?) and expect many concurrent reads of that data.

If we really lock down cursors to be single-operator on Buffers, it breaks these hypothetical use cases.

Reply:
Isn't that what readByteString() and shared segments are for though?

Reply:
Yup. Though I’m still reluctant to have APIs where ”reading operations” do mutations.

Reply:
Is it worth having a `ReadCursor` and a `WriteCursor`?